pkg/etcdutil/etcdutil.go

```diff
@@ -265,22 +274,38 @@ func TakeAndSaveFullSnapshot(ctx context.Context, client client.MaintenanceClose
 	timeTaken := time.Since(startTime)
 	logger.Infof("Total time taken by Snapshot API: %f seconds.", timeTaken.Seconds())
 
+	var snapshotData io.ReadCloser
+	snapshotTempDBPath := filepath.Join(tempDir, "db")
+	if snapshotData, err = checkFullSnapshotIntegrity(rc, snapshotTempDBPath, logger); err != nil {
+		logger.Errorf("verification of full snapshot SHA256 hash has failed: %v", err)
+		return nil, err
+	}
+	logger.Info("full snapshot SHA256 hash has been successfully verified.")
+
+	defer func() {
+		if err := os.Remove(snapshotTempDBPath); err != nil {
+			logger.Warnf("failed to remove temporary full snapshot file: %v", err)
+		}
+		if err := snapshotData.Close(); err != nil {
+			logger.Warnf("failed to close snapshot data file: %v", err)
+		}
+	}()
+
 	if cc.Enabled {
 		startTimeCompression := time.Now()
-		rc, err = compressor.CompressSnapshot(rc, cc.CompressionPolicy)
+		snapshotData, err = compressor.CompressSnapshot(snapshotData, cc.CompressionPolicy)
```

Comment:
If an error occurs here, `snapshotData` will be `nil`. The deferred `snapshotData.Close()` will thus cause a panic.

Although the code says the following error returning block is unreachable, it's better to be be on the safer side.

https://github.com/gardener/etcd-backup-restore/blob/8a3bc2c87cc32dcd4dff9d1d347531cc332638a3/pkg/compressor/compressor.go#L36-L41

I've verified that calling a `Close()` on a `nil` will panic.

```
func main() {
	a := returnNil()
	a.Close()
}

func returnNil() io.ReadCloser {
	return nil
}
```

```
➜  temp go run .
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x10269b0bc]

goroutine 1 [running]:
main.main()
        /Users/renormalize/go/src/github.com/gardener/temp/temp/main.go:9 +0x1c
exit status 2
```

Reply:
Resolved by [002c481](https://github.com/gardener/etcd-backup-restore/pull/779/commits/002c48192fd0eb97fef1ce46fb0344d4f8acbf27).